### ISSUE TITLE
Automatic path detection

### ISSUE THREAD

**Issue by Sigvaard** [fastbuild/fastbuild#511]
Fastbuild could automatically look(in C++) for certain paths and either set environment variable or set some global variables like VS2017_PATH. 

This could be done from C++ or in BFF if it supported reading registry on Windows for example

**Comment by ffulin** [comment]
It is possible to import environment variables and use those to set the path for your compiler. This can be use to obtain the installation directory of toolchains that provide it.

There has been some discussion about registry key importing here: https://github.com/fastbuild/fastbuild/issues/101

Did you have something specific in mind? Is there a particular compiler you want to detect, or a particular env var or registry key you are thinking about?

**Comment by Sigvaard** [comment]
@ffulin it would be good to detect VS paths first, versions change frequently and changing paths becomes tedious and require some maintenance. At some point projects just won't compile because of the incorrect paths.

It could be handled by some external tool that runs before fastbuild and that tool could set appropriate environment variables. But it would be greater if fastbuild alone would be capable of that. Either doing it automatically (in C++) or by giving the ability to do it in BFF. 

Doing it in BFF would be certainly more complicated to achieve due to limited language features. And sometimes finding just by registry could not be enough.

This is how CMake does it: https://github.com/Kitware/CMake/blob/master/Source/cmGlobalVisualStudio7Generator.cxx#L145

**Comment by ffulin** [comment]
@Sigvaard Thanks for the link. I tool a look at the code and it seems that this actually detects the devenv.exe location, rather than the compiler. Since CMake isn't actually a build system, this works for them and they rely on devenv finding the compiler, but that's not enough for us.

It used to be really easy to find the compiler via the global VSXXXCOMNTOOLS env vars, but since since VS2017, the compiler is in a different folder for every minor version and there is no more global env var. It is possible to import the following environment variables in the bff if you run under the "developer command prompt" though:
```C++
VCToolsInstallDir=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.15.26726\
VCToolsRedistDir=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Redist\MSVC\14.15.26706\
```

I'm definitely open to streamlining this some more, but I think for the moment, importing those env vars is probably the way to go.

Let me know if I've missed something.

**Comment by bburgin** [comment]
1. I second ffulin's suggestion to use env variables for msvc paths.
2. Please don't couple FBuild.exe to the installed Visual Studio.
    I am currently relying on FBuild.exe's ability to call cl.exe without Visual Studio being installed.
    This way, my dev machine setup is much faster:
    I can copy the cl.exe and .dll files for every msvc version that I need to a machine,
    and not have to install every Visual Studio version on the machine.

**Comment by ffulin** [comment]
All the projects I'm closely involved in store their toolchain binaries (compilers, linkers and literally everything else needed to make a build, including FASTBuild) in source control and set things up so you can compile on a brand new machine with nothing installed except the source control tool and just by syncing the depot. I recommend to anyone that will listen to vendorize everything this way. Checking in a new compiler and knowing that you've immediately updated every build machine and developer in the office to the exact same version is a truly beautiful thing.

With this in mind, FASTBuild will never by coupled to the installation of Visual Studio or any other application.

I do appreciate however, that not everyone wants to be as hardcore about these things as I do. I think it's perfectly fine to have functionality exposed that can allow someone to setup a bff to "discover" an installed compiler if that's the way they want to work. 

Hybrid approaches are also possible. For example, I plan to eventually have the FASTBuild source work this way:
 - check for a vendorized toolchain in a relative directory (how i make official builds)
 - fallback to checking for installed toolchains (for people who want to work contribute)
This will greatly simplify building FASTBuild binaries for contributors as they won't have to have the same vendorized setup.

**Comment by bburgin** [comment]
I agree. Can it be an optional field that enables the fallback logic, with the default being disabled?
This way, my builds that should not use an installed Visual Studio don't try to use the installed Visual Studio by accident.

**Comment by ffulin** [comment]
To clarify, there will be no behaviour change for existing bff files.

There might be additional functionality available to you in the bff which can be explicitly used to detect the presence of files if you want that behaviour. I would use this in FASTBuild's bff files (used when you are compiling FASTBuild itself) to make it easier for new users to build FASTBuild from source.

**Comment by bburgin** [comment]
Thanks.

**Comment by missmah** [comment]
> All the projects I'm closely involved in store their toolchain binaries (compilers, linkers and literally everything else needed to make a build, including FASTBuild) in source control and set things up so you can compile on a brand new machine with nothing installed except the source control tool and just by syncing the depot. I recommend to anyone that will listen to vendorize everything this way. Checking in a new compiler and knowing that you've immediately updated every build machine and developer in the office to the exact same version is a truly beautiful thing.

I cannot agree more with this sentiment (and would further note other benefits like coherent time travel debugging), and it's been truly bizarre realizing that some very large companies don't already do this / are hesitant to do this / actually install build tools onto hundreds or thousands of build machines.

> With this in mind, FASTBuild will never by coupled to the installation of Visual Studio or any other application.

<3
